### PR TITLE
t/23: Introduced the Plugin loading API

### LINKED ISSUE #23: Introducing the Plugins API
Introduce the Plugins API according to the design docs:
https://github.com/ckeditor/ckeditor5-design/wiki/Plugins

### PR DESCRIPTION
**Pull request by fredck**
Fixes #23.

Note that relative updates ended up in [ckeditor5](https://github.com/ckeditor/ckeditor5), where a new test plugin has been introduced [ckeditor5-plugin-devtest](https://github.com/ckeditor/ckeditor5-plugin-devtest). Therefore `git pull` and `npm install` must be executed in `ckeditor5` to catch these updates.

### CODE REVIEW

**Comment by fredck** [comment]
Right after creating this PR, one thing came to my mind... the way it is coded now (using Promise.all to wait for all init() calls to resolve) may not guarantee the initialization order if a asynchronous plugin depends on a non-asynchronous one.

I'll be checking this, so this aspect should be ignored on review, but still a R+ should not take place until we'll not have this issue clarified. I'll follow up on this PR about this.

**Comment by fredck** [comment]
> Right after creating this PR, one thing came to my mind... the way it is coded now (using Promise.all to wait for all init() calls to resolve) may not guarantee the initialization order if a asynchronous plugin depends on a non-asynchronous one.

In fact I was right with my doubt. I've pushed an additional commit to solve this issue. Ready for full review.

**Comment by fredck** [comment]
Updated the PR description with additional information.

File: src/editor.js [hunk 1]
```diff
@@ -42,6 +42,63 @@ CKEDITOR.define( [ 'mvc/model', 'editorconfig' ], function( Model, EditorConfig
...
+		 * directly.
+		 *
+		 * @returns {Promise} A promise which resolves once the initialization is completed.
+		 */
+		init: function() {
+			var that = this;
+			var config = this.config;
+
+			// Create and cache a promise that resolves when all initialization procedures get resolved.
+			this._initPromise = this._initPromise || Promise.all( [
```
Comment (gregpabian) [inline src/editor.js]: Shouldn't you require the "promise" module instead of using global Promise?

File: src/plugincollection.js [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+		/**
+		 * Gets a plugin from the collection.
+		 *
+		 * @param {String} name The plugin name.
+		 * @returns {Plugin} The requested plugin, if available in the collection.
+		 */
+		get: function( name ) {
+			if ( typeof name != 'string' ) {
+				return Collection.prototype.get.apply( this, arguments );
```
Comment (gregpabian) [inline src/plugincollection.js]: Do we really need this?

File: src/plugincollection.js [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+			// The hash table used to store pointers to loaded plugins by name.
+			this._names = {};
+		},
+
+		/**
+		 * Loads a set of plugins and add them to the collection.
+		 *
+		 * @param {String} plugins A comma separated list of plugin names to get loaded.
+		 */
+		load: function( plugins ) {
```
Comment (gregpabian) [inline src/plugincollection.js]: Missing documentation for the return value.

File: tests/editor/editor.js [hunk 4]
```diff
@@ -21,6 +21,60 @@ beforeEach( function() {
...
+		init: sinon.spy().named( 'F' )
+	} );
+} );
+
+var asyncSpy = sinon.spy().named( 'async-call-spy' );
+
+CKEDITOR.define( 'plugin!async', [ 'plugin' ], function( Plugin ) {
+	return Plugin.extend( {
+		init: sinon.spy( function() {
+			return new Promise( function( resolve ) {
```
Comment (gregpabian) [inline tests/editor/editor.js]: You should require the "promise" module instead of using the global one.

File: src/plugincollection.js (hunk 2 above)
Comment (gregpabian) [inline src/plugincollection.js]: Ok, I saw you use this case in the tests, but is there any other practical use for this?

File: src/editor.js [hunk 5]
```diff
@@ -42,6 +42,63 @@ CKEDITOR.define( [ 'mvc/model', 'editorconfig' ], function( Model, EditorConfig
...
+				}
+
+				// Return the promise chain.
+				return promise;
+
+				function getInitResolveFn( index ) {
+					return function() {
+						// Resolve with the return value of init(). If it is a Promise, it'll inherit its state. For
+						// everything else it resolves immediately.
+						return Promise.resolve( that.plugins.get( index ).init() );
```
Comment (gregpabian) [inline src/editor.js]: Don't know why but my previous comment somehow disappeared... Anyway, I think You don't have to wrap the returned value in a promise again. This should work just fine:

```js
return that.plugins.get( index ).init();
```

File: src/plugincollection.js (hunk 2 above)
Comment (fredck) [inline src/plugincollection.js]: You mean, the ability to retrieve a plugin by name? Yes this is very important when one wants to interacts with the plugin instance in the editor directly, for example calling methods of that plugin.

Comment (gregpabian) [inline src/plugincollection.js]: No, I mean the ability to retrieve a plugin by index.

Comment (fredck) [inline src/plugincollection.js]: Ah, then yes, definitely. That's one of the basic features in collections so one can iterate through the items in a loop. (e.g. listing all plugins).

Comment (gregpabian) [inline src/plugincollection.js]: OK, makes sense.

File: src/plugincollection.js (hunk 3 above)
Comment (fredck) [inline src/plugincollection.js]: Fixed with de01ab4.

File: src/editor.js (hunk 1 above)
Comment (fredck) [inline src/editor.js]: Definitely! Fixed with 93ac426. I've additionally removed `Promise` from the JSHint globals with cb00b6e so it'll now throw errors in such cases.

File: tests/editor/editor.js (hunk 4 above)
Comment (fredck) [inline tests/editor/editor.js]: Fixed with 93ac426.

File: src/editor.js (hunk 5 above)
Comment (fredck) [inline src/editor.js]: Good one... there is not reason to have all `.then()` calls in the chain returning promises.

Fixed with 216a50a.

File: src/editor.js (hunk 1 above)
Comment (gregpabian) [inline src/editor.js]: > I've additionally removed Promise from the JSHint globals

+1 I was about to suggest it :)

**Comment by fredck** [comment]
Ok, all issues fixed. Awaiting for review again.

**Pull request merged**